datafusion/physical-expr/src/physical_expr.rs

```diff
@@ -152,124 +283,82 @@ pub type PhysicalExprRef = Arc<dyn PhysicalExpr>;
 /// the boundaries for all known columns.
 #[derive(Clone, Debug, PartialEq)]
 pub struct AnalysisContext {
-    /// A list of known column boundaries, ordered by the index
-    /// of the column in the current schema.
-    pub column_boundaries: Vec<Option<ExprBoundaries>>,
-    // Result of the current analysis.
-    pub boundaries: Option<ExprBoundaries>,
+    // A list of known column boundaries, ordered by the index
+    // of the column in the current schema.
+    pub boundaries: Option<Vec<ExprBoundaries>>,
+    /// The estimated percentage of rows that this expression would select, if
+    /// it were to be used as a boolean predicate on a filter. The value will be
+    /// between 0.0 (selects nothing) and 1.0 (selects everything).
+    pub selectivity: Option<f64>,
 }
 
 impl AnalysisContext {
-    pub fn new(
-        input_schema: &Schema,
-        column_boundaries: Vec<Option<ExprBoundaries>>,
-    ) -> Self {
-        assert_eq!(input_schema.fields().len(), column_boundaries.len());
+    pub fn new(boundaries: Vec<ExprBoundaries>) -> Self {
         Self {
-            column_boundaries,
-            boundaries: None,
+            boundaries: Some(boundaries),
+            selectivity: None,
         }
     }
 
-    /// Create a new analysis context from column statistics.
-    pub fn from_statistics(input_schema: &Schema, statistics: &Statistics) -> Self {
-        // Even if the underlying statistics object doesn't have any column level statistics,
-        // we can still create an analysis context with the same number of columns and see whether
-        // we can infer it during the way.
-        let column_boundaries = match &statistics.column_statistics {
-            Some(columns) => columns
-                .iter()
-                .map(ExprBoundaries::from_column)
-                .collect::<Vec<_>>(),
-            None => vec![None; input_schema.fields().len()],
-        };
-        Self::new(input_schema, column_boundaries)
-    }
-
-    pub fn boundaries(&self) -> Option<&ExprBoundaries> {
-        self.boundaries.as_ref()
-    }
-
-    /// Set the result of the current analysis.
-    pub fn with_boundaries(mut self, result: Option<ExprBoundaries>) -> Self {
-        self.boundaries = result;
-        self
+    pub fn new_with_selectivity(
+        boundaries: Vec<ExprBoundaries>,
+        selectivity: f64,
+    ) -> Self {
+        Self {
+            boundaries: Some(boundaries),
+            selectivity: Some(selectivity),
+        }
```

Comment:
An alternate form that might be more pleasing to the eye might be 

```suggestion
    pub fn with_selectivity(
        mut self,
        selectivity: f64,
    ) -> Self {
        self.selectivity = Some(selectivity);
        self
       }
```

then instead of 

```rust
    Ok(AnalysisContext::new_with_selectivity(
        target_boundaries,
        selectivity,
    ))
```

you could write

```rust
    Ok(AnalysisContext::new(target_boundaries)
        .with_selectivity(selectivity)
    )
```